### PR TITLE
task/FP-1960: Edit registrations action

### PR DESCRIPTION
**Pull request by edmondsgarrett**
## Overview
Added action on registrations listing admin page for admins to edit registration records and write edits to the database
…

## Related

- [FP-1960](https://jira.tacc.utexas.edu/browse/FP-1960)
…

## Changes
- Added template for edit registration modal (mimics the listing registration page, but all fields have default values from the submitted registration request)
- Updated `views.py` to catch multiple POST forms (this and create submitter), as well as some data formatting for convenience
- Added db util functions to update registration/entities/contacts, or delete entities/contacts (also updated some of the create db actions for synergy between these new functions)
- Updated listing page to add new action, as well as other action templates based on some updates to `views.py`
…

## Testing

*** At this time, testing database read/write locally is still not possible, so expect any attempts to submit this action to result in the error page ***
1. Replace lines 13-15 in `apps/admin_regis_table/views.py` with the following:
   <details><summary>Snippet</summary>
     
          import datetime
          registrations_content = [
            (
                1,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                False,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'received',                           #registration_status
                'insurance carrier',                #org_type
                'Golden Rule Insurance Company',    #business_name
                '7440 Woodland Drive',              #mail_address
                'Indianpolis',                      #city
                'IN',                               #state
                '46278     '                        #zip
            ),
            (
                2,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                False,                               #submitting_for_self
                'SFTP',                             #submission_method
                'processing',                           #registration_status
                'pbm',                #org_type
                'Clay Tenant Insurance Company',    #business_name
                '440 Wood and Drive',              #mail_address
                'Indianapolis',                      #city
                'NY',                               #state
                '24444     '                        #zip
            ),
            (
                3,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'complete',                        #registration_status
                'insurance carrier',                #org_type
                'Magenta Condition Insurance Company',    #business_name
                '742 Oddland Drive',              #mail_address
                'Indipolis',                      #city
                'MI',                               #state
                '62784     '                        #zip
            )
        ]
        registrations_entities = [
            (
                5, 1, 5, 46, 11111, 10001, 5, 'Garretts Test Business 2', '11-0001111'
            ),
            (
                5, 2, 50000, 47, 11010, 21101, 1, 'A Second Test Entity', '00-0000000'
            )
        ]
        registrations_contacts = [
            (
                52,
                1,
                False,
                'Test Role',
                'Garrett Test Tester',
                '2222222222',
                'notarealemail@emailplace.email'
            ),
            (
                53,
                1,
                False,
                'A 2nd Test Role',
                'Test Garrett 2 Test',
                '15555555555',
                'testemail@testemail.emailtest'
            ),
            (
                54,
                2,
                True,
                'A 3rd and final test role',
                'Test 3rd Role Name Garrett',
                '0000000000',
                None
            )
          ]
        
    </details>
2. Open http://localhost:8000/administration/list-registration-requests/
3. Open any of the Edit Registration links in the Actions column of the table
4. In another tab, open https://apcd-qa.tacc.utexas.edu/register/request-to-submit/
5. Compare the Edit Registration modal to the registration form: look, functionality, etc.
6. Ensure that changes cannot be submitted without all required fields having a value, or that changes cannot be submitted on certain fields if the format is wrong
7. Try submitting changes at least once, and ensure you are directed to an error page upon submission

## UI

…

### CODE REVIEW

**Review by shayanaijaz — CHANGES_REQUESTED** [review]
Found a bug with the `NAIC Company Code` validation in the Edit Record modal. The validation does not happen on that field **if the field is left untouched and there was a previous error in the field** (from the sample test data). If I edit the field and then submit then the proper error comes up

**Comment by edmondsgarrett** [comment]
> Found a bug with the `NAIC Company Code` validation in the Edit Record modal. The validation does not happen on that field **if the field is left untouched and there was a previous error in the field** (from the sample test data). If I edit the field and then submit then the proper error comes up


I see it! I believe this is related to how NAIC codes with leading `0`'s are written to the database - values are set as `int` before being written, so any leading zeroes would be stripped away. I think this is a catch to add to the registration form as well as this form.

File: apcd-cms/src/apps/admin_regis_table/templates/create_submitter_modal.html [hunk 1]
```diff
@@ -39,11 +39,11 @@ <h4>File Submission</h4>
           </article>
         </div>
 
-        <form method="POST" action="" class="c-form">
+        <form method="POST" action="" class="c-form" name="create-submitter-form">
           <h4>Submitter Data</h4>
 
           {% csrf_token %}
-          <input type="hidden" name="_token" value>
+          <input type="hidden" name="create-submitter-form" value>
```
Comment (wesleyboar) [inline apcd-cms/src/apps/admin_regis_table/templates/create_submitter_modal.html:46]: Is this intentional?

File: apcd-cms/src/apps/admin_regis_table/templates/edit_registration_modal.html [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+    );
+    const inputListener = e => {
+      inputs.filter(i => i !== e.target)
+            .forEach(i => (i.required = !e.target.value.length));
+    };
+
+    inputs.forEach(i => i.addEventListener('input', inputListener));
+  </script>
+</aside>
+
```
Comment (wesleyboar) [inline apcd-cms/src/apps/admin_regis_table/templates/edit_registration_modal.html:889]: It is not easily maintainable to have two forms so similar. Please try to create one template to share as much as feasible for both forms.

**Review by wesleyboar — CHANGES_REQUESTED** [review]

File: apcd-cms/src/apps/admin_regis_table/templates/edit_registration_modal.html (hunk 2 above)
Comment (edmondsgarrett) [reply apcd-cms/src/apps/admin_regis_table/templates/edit_registration_modal.html]: Very good point, just pushed up changes to reflect this feedback!

**Review by wesleyboar — APPROVED** [review]
Approved based **solely** on big picture template changes. I beg another reviewer to test functionality. I am sorry for not being more available.

**Review by shayanaijaz — APPROVED** [review]
LGTM!

File: apcd-cms/src/apps/admin_regis_table/templates/create_submitter_modal.html (hunk 1 above)
Comment (edmondsgarrett) [reply apcd-cms/src/apps/admin_regis_table/templates/create_submitter_modal.html:46]: Yep! I am using this to differentiate between different POST forms on this page.

**Pull request merged**